ECAM Cruise page

Jargendas opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

Hi, what a great project!

**Summary of Changes**

I started implementing an ECAM **Cruise** page as I noticed it does not seem to be worked on at the moment. I mainly used some images from the internet as reference ([here](https://aviationmatters.co/what-are-eicas-and-ecam-systems-on-aircraft/) and [here](https://forums.x-plane.org/index.php?/forums/topic/184581-some-minor-ecam-display-issues/)) as well as Aerosoft's A320 for FSX. It definitely needs a bit of polishing (like the value alignments) and feedback from people who know more about the A320's systems than me.
Furthermore, I have not been able to properly integrate it as in real life yet, so I replaced the HYD page for testing. But I am fairly certain that a better integration is possible without high effort.

All values that were currently available are implemented and dispayed correctly as far as I can tell. Cabin landing elevation and cabin temperatures cannot be modified in the model (yet) and are therefore fixed values.

I am looking forward to feedback and hope this is useful for the project!

**Screenshots**

![cruise-page-screenshot](https://user-images.githubusercontent.com/1270412/92316340-b241c400-eff2-11ea-9afa-31091ff54b6b.PNG)

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_CRZ.html:
```diff
@@ -0,0 +1,97 @@
+            <text class="Unit" x="320" y="450">°C</text> 
+
+    </svg>
+<!--/div-->
```
Same applies.

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_CRZ.js:
```diff
@@ -0,0 +1,122 @@
+            this.ForwardTemp = this.querySelector("#ForwardTemp");
+            this.AftTemp = this.querySelector("#AftTemp");
+
+            //this.querySelector("#LandingElevationMode").textValue 
```
Clairfy or remove comment.

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_CRZ.js:
```diff
@@ -0,0 +1,122 @@
+            
+            // Fuel
+            var unitFactor = SimVar.GetSimVarValue("FUEL WEIGHT PER GALLON", "lbs");
+            if (BaseAirliners.unitIsMetric(Aircraft.A320_NEO))
```
We always use brackets for if statements. Sorry, eslint config not ready yet

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_CRZ.js:
```diff
@@ -0,0 +1,122 @@
+            this.FuelUsedRight.textContent = fastToFixed(rightConsumption, 0);
+
+            // Oil
+            var value = SimVar.GetSimVarValue("ENG OIL QUANTITY:1", "percent") * 0.01 * 25;
```
Use `let`.

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_CRZ.js:
```diff
@@ -0,0 +1,122 @@
+
+            // Engines
+            value = SimVar.GetSimVarValue("TURB ENG VIBRATION:1", "Number");
+            if (value < 0) value = 0.0;
```
note: those are fine, as they're very short.

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_CRZ.js:
```diff
@@ -0,0 +1,122 @@
+            value = SimVar.GetSimVarValue("PRESSURIZATION CABIN ALTITUDE", "feet");
+            this.CabinAltitude.textContent = fastToFixed(value, 0);
+
+            var cabinAltMetres = value * 0.3048;
+            var cabinPressurePascal = 101325 * Math.exp(-0.00011857591*cabinAltMetres);
+            var cabinPressurePSI = cabinPressurePascal * 0.000145038;
+            var outsidePressureINHG = SimVar.GetSimVarValue("AMBIENT PRESSURE", "inHg");
+            var outsidePressurePSI = outsidePressureINHG * 0.491154;
+            var pressureDiff = cabinPressurePSI - outsidePressurePSI;
```
Use `const` as those do not change, and fix spacing in `Math.exp(...)` at line 107

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_CRZ.js:
```diff
@@ -0,0 +1,122 @@
+            TemplateElement.call(this, this.init.bind(this));
+        }
+        init() {
+            this.FuelUsedTotal = this.querySelector("#FuelUsedTotal"); //
+            this.FuelUsedLeft = this.querySelector("#FuelUsedLeft"); //
+            this.FuelUsedRight = this.querySelector("#FuelUsedRight"); //
+
+            this.OilQuantityLeft = this.querySelector("#OilQuantityLeft"); //
+            this.OilQuantityLeftDecimal = this.querySelector("#OilQuantityLeftDecimal");
+            this.OilQuantityRight = this.querySelector("#OilQuantityRight"); //
+            this.OilQuantityRightDecimal = this.querySelector("#OilQuantityRightDecimal");
+
+            this.VibN1Left = this.querySelector("#VibN1Left"); //
+            this.VibN1LeftDecimal = this.querySelector("#VibN1LeftDecimal");
+            this.VibN1Right = this.querySelector("#VibN1Right"); //
+            this.VibN1RightDecimal = this.querySelector("#VibN1RightDecimal");
+
+            this.VibN2Left = this.querySelector("#VibN2Left"); //
+            this.VibN2LeftDecimal = this.querySelector("#VibN2LeftDecimal");
+            this.VibN2Right = this.querySelector("#VibN2Right"); //
+            this.VibN2RightDecimal = this.querySelector("#VibN2RightDecimal");
+
+            this.LandingElevationMode = this.querySelector("#LandingElevationMode");
+            this.LandingElevation = this.querySelector("#LandingElevation");
+
+            this.DeltaPressure = this.querySelector("#DeltaPressure");
+            this.DeltaPressureDecimal = this.querySelector("#DeltaPressureDecimal");
+
+            this.CabinVerticalSpeed = this.querySelector("#CabinVerticalSpeed"); //
+            this.CabinAltitude = this.querySelector("#CabinAltitude"); //
```
Why the `//` at the end of those lines ?

Jargendas replied:
I used them to remember which of those I already integrated in the JS, forgot to remove afterwards.

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_LowerECAM_CRZ.html:
```diff
@@ -0,0 +1,97 @@
+<link rel="stylesheet" href="./A320_Neo_LowerECAM_CRZ.css" />
+
+<script type="text/html" id="LowerECAMCRZTemplate">
+<!--div id="a320-neo-lower-ecam-crz" style="width: 600px; height: 600px; background-color: black;"-->
```
Clarify reason of commented code or remove it entirely

Jargendas replied:
Was just for my testing, so I could view it as a website in a browser. I will remove it later.

Benjozork reviewed (commented):
Wow ! This is genuinely a great PR. Thanks for the detailed information and clean code. I added some notes about formatting in my review but at first glance this seems solid.

We will defnitely be starting the process of reviewing this with IRL A320 pilots (if you are not in our discord, I invite you to join as that is where we have those discussions. However, we do add the gathered info to GitHub afterwards when possible). I hope we can work with you to get this up to completion and hopefully merge this in !

One question though, was this made in coordination with #440 ?

Jargendas commented:
Great, I will definitely join the Discord and continue improving this.
And no, I don't have anything to do with #440, that seems to be just a coincidence. :)

Jargendas commented:
Okay, this is pretty done now from my point of view and can be shown to some IRL pilots. :)
I removed the bad value alignments and added a few other improvements.

I also integrated it properly like the real cruise page (572039d): When you press an already illuminated button, the button light turns off and the cruise page shows up. However, to do so, I had to include a modified version of BaseAirliners.js (from `asobo-vcockpits-instruments-airliners`) to the repo, not sure if that is okay. Only `BaseEICAS.changePage()` and the constructor of `BaseEICAS` are modified in BaseAirliners.js. I added a flag to the constructor so that my modifications should only affect the A320, but it's still a very clumsy solution.
I initially thought it should be possible - and more elegant - to just override the callbacks from BaseEICAS in A320_Neo_EICAS, but I wasn't able to pull it off. My JavaScript is not very good though, so maybe someone else is able to do it.

Updated screenshot of how it looks now:

<img src="https://user-images.githubusercontent.com/1270412/92336881-9d723880-f0a5-11ea-836a-709ec7d67ced.PNG" width="400">

wpine215 commented:
Looks good to me, but we should really implement some sort of method overriding in A320_Neo_EICAS, since we really don't want to touch shared classes, since updates can break those and cause a mess. Should hold off on merging until that's implemented.

Jargendas commented:
> 
> 
> Looks good to me, but we should really implement some sort of method overriding in A320_Neo_EICAS, since we really don't want to touch shared classes, since updates can break those and cause a mess. Should hold off on merging until that's implemented.

Yeah, that's what I thought. I'll try again and look for help in the discord.

Jargendas commented:
I was pointed to [msfs-webui-devkit](https://github.com/dga711/msfs-webui-devkit), which made actual debugging possible. With this, the problem was easy to solve!

Jargendas commented:
@wpine215 Soo - merge?
👉👈

Benjozork commented:
@Jargendas we usually test every PR individually since automated testing is just not possible - and verify with real pilots. This process may take time, sorry.

What is your discord username btw ?

Jargendas commented:
Ah okay, sorry. Discord name should be Jargendas as well.

Armankir commented:
Tested, the page did not automatically come up after flaps retract and gear up. Not a real pilot and would like someone to double-check if I did something incorrectly, but would like to let you know.

wpine215 commented:
Works perfectly for me now, merging.
